### PR TITLE
keep ResourceBinding when removing PropagationPolicy

### LINKED ISSUE #505: Do not delete the work when PropagationPolicy is deleted
**What would you like to be added**:
Do  not delete the work in member clusters when the related propagationPolicy is deleted
**Why is this needed**:
The users may delete the propagationPolicy (or even the crd) and then they may find that the all the related works in member are deleted. It may be unexpected 。

So I think, that the work may not be deleted when the  propagationPolicy is missing， and will be deleted  when the workload itself  is deleted.

### PR DESCRIPTION
**Pull request by qianjun1993**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

stop delete resourcebinding when the related policy is deleted 

**Which issue(s) this PR fixes**:
Fixes #505

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by RainbowMango** [comment]
I haven't dig into why the E2E failing. Just echo the logs here and try again.

```
Export Service from source-clusters, import Service to destination-clusters
  /root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:346
STEP: Create ClusterPropagationPolicy(serviceexports-d92-policy) to Propagation ServiceExport CRD
STEP: Create ClusterPropagationPolicy(serviceimports-hwh-policy) to Propagation ServiceImport CRD
STEP: Wait ServiceExport CRD present on member clusters
I0813 15:16:46.647139  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:16:51.663366  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:16:56.664463  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:01.663626  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:06.663848  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:11.663597  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:16.663025  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:21.693098  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:26.663388  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:31.663187  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:36.663010  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:41.662953  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:46.670273  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:51.662679  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:17:56.662566  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:18:01.667378  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:18:06.662612  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
I0813 15:18:11.665766  598021 mcs_test.go:224] Waiting for CRD(ServiceExport) present on member cluster(member1)
STEP: Wait ServiceImport CRD present on member clusters
STEP: Create Deployment(karmadatest-5dl/hello-vp2) in member1 cluster
STEP: Create Service(karmadatest-5dl/hello-vp2) in member1 cluster
STEP: Wait Service(karmadatest-5dl/hello-vp2)'s EndpointSlice exist in member1 cluster
STEP: Create ServiceExport(karmadatest-5dl/hello-vp2)
STEP: Create PropagationPolicy(karmadatest-5dl/export-hello-vp2-policy) to propagate ServiceExport
STEP: Wait EndpointSlices collected to namespace(karmadatest-5dl) in controller-plane
STEP: Create ServiceImport(karmadatest-5dl/hello-vp2)
STEP: Create PropagationPolicy(karmadatest-5dl/import-hello-vp2-policy) to propagate ServiveImport
STEP: Wait derived-service(karmadatest-5dl/derived-hello-vp2) exist in member2 cluster
STEP: Wait EndpointSlices have been imported to member2 cluster
STEP: TCP connects across clusters using the ClusterIP
STEP: UDP connects across clusters using the ClusterIP
STEP: Cleanup
STEP: Delete Deployment(karmadatest-5dl/hello-vp2) in member1 cluster
STEP: Delete Service(karmadatest-5dl/hello-vp2) in member1 cluster
STEP: Delete ClusterPropagationPolicy serviceexports-d92-policy
STEP: Delete ClusterPropagationPolicy serviceimports-hwh-policy
STEP: Wait ServiceExport CRD disappear on member clusters

• Failure in Spec Teardown (AfterEach) [430.464 seconds]
[MCS] Multi-Cluster Service testing
/root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:178
  Connectivity testing [AfterEach]
  /root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:298
    Export Service from source-clusters, import Service to destination-clusters
    /root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:346

    Unexpected error:
        <*errors.errorString | 0xc0002227d0>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    occurred

    /root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:278
------------------------------
SSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] [MCS] Multi-Cluster Service testing [AfterEach] Connectivity testing Export Service from source-clusters, import Service to destination-clusters 
/root/actions-runner/_work/karmada/karmada/test/e2e/mcs_test.go:278

Ran 6 of 22 Specs in 466.684 seconds
FAIL! -- 5 Passed | 1 Failed | 0 Pending | 16 Skipped
--- FAIL: TestE2E (466.80s)
FAIL

```

**Comment by qianjun1993** [comment]
Do you know when the crd of ServiceImport CRD is created. Is it created in advance?  I only found the creation of ClusterPropagationPolicy in this test.

**Comment by XiShanYongYe-Chang** [comment]
> Do you know when the crd of ServiceImport CRD is created. Is it created in advance? I only found the creation of ClusterPropagationPolicy in this test.

The `ServiceImport` and `ServiceExport` CRD were created in karmada installation:

https://github.com/karmada-io/karmada/blob/51233e54088bb3ca71b8ae1d058de6e9fb99c232/hack/deploy-karmada.sh#L115-L116

**Comment by XiShanYongYe-Chang** [comment]
You may need to modify the e2e test:
https://github.com/karmada-io/karmada/blob/51233e54088bb3ca71b8ae1d058de6e9fb99c232/test/e2e/mcs_test.go#L265-L296

the original logic will check whether `ServiceExport` and `ServiceImport` CRD has been deleted in member clusters.

**Comment by qianjun1993** [comment]
So just remove this check?

**Comment by XiShanYongYe-Chang** [comment]
> So just remove this check?

I think so.

**Comment by RainbowMango** [comment]
/assign
I'll review it ASAP, thanks.

File: test/e2e/mcs_test.go [hunk 1]
```diff
@@ -258,41 +258,6 @@ var _ = ginkgo.Describe("[MCS] Multi-Cluster Service testing", func() {
 			err := karmadaClient.PolicyV1alpha1().ClusterPropagationPolicies().Delete(context.TODO(), serviceImportPolicy.Name, metav1.DeleteOptions{})
 			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
 		})
```
Comment (XiShanYongYe-Chang) [inline test/e2e/mcs_test.go:260]: Can you add some comment to record why not wait `ServiceExport` and `ServiceImport` disappear on member clusters?

Comment (qianjun1993) [reply test/e2e/mcs_test.go:260]: OK, I will add it.

File: pkg/util/detector/detector.go [hunk 2]
```diff
@@ -894,48 +888,36 @@ func (d *ResourceDetector) HandleClusterPropagationPolicyDeletion(policyName str
 		policyv1alpha1.ClusterPropagationPolicyLabel: policyName,
 	}
 
-	// load and remove the ClusterResourceBindings which labeled with current policy
+	// load the ClusterResourceBindings which labeled with current policy
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:891]: The method's comments should be updated as well, e.g.:
```golang
// HandleClusterPropagationPolicyDeletion handles ClusterPropagationPolicy delete event.
// After a policy is removed, the label marked on relevant resource template will be removed(which gives
// the resource template a change to match another policy).
//
// Note: The relevant ClusterResourceBinding or ResourceBinding will continue to exist until the resource template is gone.
func (d *ResourceDetector) HandleClusterPropagationPolicyDeletion(policyName string) error {
```

**Comment by RainbowMango** [comment]
I'm considering a case,

- S0: Suppose we have a PropagationPolicy `a` which propagates a deployment to `member1`.
- S1: Remove PropagationPolicy `a`; 
- S2: Create PropagationPolicy `b` which propagates the same deployment to `member2`.

Will the deployment be removed from `member1`? If not, does the behavior meets our expectation?

File: pkg/util/detector/detector.go [hunk 3]
```diff
@@ -866,20 +866,14 @@ func (d *ResourceDetector) HandlePropagationPolicyDeletion(policyNS string, poli
 		return err
 	}
 
-	for itemIndex, binding := range rbs.Items {
+	for _, binding := range rbs.Items {
```
[the remark this answers is not in the document]
Comment (qianjun1993) [reply pkg/util/detector/detector.go:870]: OK, I will change it.

**Comment by qianjun1993** [comment]
> I'm considering a case,
> 
> * S0: Suppose we have a PropagationPolicy `a` which propagates a deployment to `member1`.
> * S1: Remove PropagationPolicy `a`;
> * S2: Create PropagationPolicy `b` which propagates the same deployment to `member2`.
> 
> Will the deployment be removed from `member1`? If not, does the behavior meets our expectation?

The deployment will be removed from  member1

**Comment by RainbowMango** [comment]
I found this issue twice today. Let's re-trigger it.
```
Installing 'kubectl v1.18.0' for you, may require the root privileges
curl: (56) TCP connection reset by peer
Error: Process completed with exit code 1.
```

**Comment by RainbowMango** [comment]
/approve
Thanks.

**Pull request merged**